tools/please_pex/preamble/path.c

```diff
@@ -0,0 +1,234 @@
+    if (sysctl(mib, NELEMS(mib), exe_path, &len, NULL, 0) == -1) {
+        err = err_from_errno("sysctl write");
+        goto end;
+    }
+#else
+#error "Unsupported operating system"
+#endif
+
+    exe_dir = dirname(exe_path);
+    if (((*pex_dir) = strdup(exe_dir)) == NULL) {
+        err = err_from_errno("strdup");
+        goto end;
+    }
+
+end:
+#ifdef __APPLE__
+    FREE(exe_path);
+#endif
+
+    return err;
+}
+
+/*
+ * get_plz_bin_path stores the path to the directory containing Please's binary outputs in path. It
+ * returns NULL on success and an error on failure.
+ *
+ * If the .pex file is running from within a Please build environment (which is assumed if PLZ_ENV
+ * is defined in the environment), get_plz_bin_path returns the path to the build environment's root
+ * directory (which is assumed to be the value of TMP_DIR in the environment, as created by Please).
+ * Otherwise, get_plz_bin_path returns the path to the bin/ subdirectory within the plz-out/ directory
+ * to which this .pex file belongs. If the .pex file does not exist within a plz-out/ directory
+ * tree, get_plz_bin_path fails.
+ */
+err_t *get_plz_bin_path(char **path) {
+    char *pex_dir = NULL;
+    char *tmp_dir = NULL;
+    char *pex_realpath = NULL;
+    char *tmp_realpath = NULL;
+    size_t pex_len = 0;
+    size_t tmp_len = 0;
+    err_t *err = NULL;
+
+    if ((err = get_pex_dir(&pex_dir)) != NULL) {
+        err = err_wrap("get_pex_dir", err);
+        goto end;
+    }
+
+    if (getenv("PLZ_ENV") != NULL) {
+        if ((tmp_dir = getenv("TMP_DIR")) == NULL) {
+            // There are two circumstances under which this might happen: either if this is a Please
+            // build environment but TMP_DIR was unset before the .pex file was executed, or if this
+            // isn't a Please build environment and PLZ_ENV just happened to be set for some other
+            // reason. The latter is more likely than the former (many build definitions rely on
+            // TMP_DIR being defined, which would lead to much more breakage besides this function
+            // failing) - therefore, optimise for the latter case.
+            log_warn("PLZ_ENV is defined but TMP_DIR is not; assuming this is not a Please build environment");
+            goto no_plz_env;
+        }
+
+        // Identify whether the .pex file exists inside the build environment.
+        if ((pex_realpath = realpath(pex_dir, NULL)) == NULL) {
+            err = err_from_errno("realpath pex_dir");
+            goto end;
+        }
+        if ((tmp_realpath = realpath(tmp_dir, NULL)) == NULL) {
+            err = err_from_errno("realpath tmp_dir");
+            goto end;
+        }
+
+        pex_len = strlen(pex_realpath);
+        tmp_len = strlen(tmp_realpath);
+
+        if (
+            strncmp(tmp_realpath, pex_realpath, tmp_len) == 0 &&
+            (
+                (pex_len == tmp_len) ||
+                (pex_len - tmp_len >= 1 && pex_realpath[tmp_len] == '/')
```

Comment:
Is there a constant defined somewhere for the system path separator?

Reply:
Sadly not. It's not the end of the world though, given that we only support platforms that use `/` (and even most Windows API functions will do the right thing if you use `/` instead of `\`).